code/ai/aicode.cpp

```diff
@@ -14646,7 +14652,41 @@ int aas_1(object *objp, ai_info *aip, vec3d *safe_pos)
 			}
 		}
 
-		if (!pos_set) {
+		if (pos_set) {
+			// if the object is not a small ship, a surface impact is likely to be some distance away from the ship center, so refine the position
+			if (fix_bugs && target_ship_obj && !Ship_info[Ships[target_ship_obj->instance].ship_info_index].is_small_ship()) {
+				// set up cache if needed
+				if (weaponp->homing_cache_ptr == nullptr) {
+					weaponp->homing_cache_ptr.reset(new homing_cache_info{ TIMESTAMP::immediate(), expected_pos, false });
+				}
+
+				// only refine the position every so often though
+				if (timestamp_elapsed(weaponp->homing_cache_ptr->next_update)) {
+					mc_info mc;
+					mc.model_instance_num = Ships[target_ship_obj->instance].model_instance_num;
+					mc.model_num = Ship_info[Ships[target_ship_obj->instance].ship_info_index].model_num;
+					mc.orient = &target_ship_obj->orient;
+					mc.pos = &target_ship_obj->pos;
+					mc.p0 = &weapon_objp->pos;				// Point 1 of ray to check
+					mc.p1 = &expected_pos;					// Point 2 of ray to check
+					mc.flags = MC_CHECK_MODEL;
+
+					model_collide(&mc);
```

Comment:
I'm sure this is fine, but it does still get my hackles up a bit, if `homing_pos` is valid for AI *and* valid when the player is targeting a subsystem (which is overwhelmingly going to be the case, since targeting subsystems is the default), it's going to be true in the vast majority of cases so it really feels like we should be using it, and saving this for when all of that isn't true, if at all.

Reply:
Valid point.  I'll see about tweaking this.

Reply:
Ok, tweaked.